fondant/component.py

```diff
@@ -153,6 +158,10 @@ def _process_dataset(self, dataset: FondantDataset) -> FondantDataset:
         """
         df = dataset.load_dataframe(self.spec)
         df = self.transform(args=self.args, dataframe=df)
+        # evolve manifest
+        dataset.manifest = dataset.manifest.evolve(self.spec)
```

Comment:
I would rather remove all manifest manipulation in the `Dataset` completely, use the manifest evolution instead, and validate the output data against the evolved manifest. So regard the manifest as immutable in the `Dataset` class.

It's also not completely clear yet how we want to leverage the `Dataset` class. Currently we mutate it, but it might be interesting to have it immutable and just use it to translate between the user dataframe and the stored data.
stored data -> `Dataset(manifest)` -> dataframe
dataframe -> `Dataset(manifest)` -> stored data

Reply:
So that means we would have to create 2 Dataset instances for transform components?
- one to wrap the input manifest and load the data from it to provide to the user
- one to wrap the output manifest and upload data to the cloud

Reply:
Yes. I just think the mutable part is done by:
- The user manipulating the dataframe
- The manifest evolution

So I don't think we need it in the dataset as well. If we follow the proposed approach, the `Dataset` class has a clear goal, which is translating between stored data and the user dataframe based on the manifest. This can include things such as data and type validation as well.

Reply:
CC: @PhilippeMoussalli @GeorgesLorre since this is about higher level design.

Reply:
> The Dataset class should be able to check the output subsets of its manifest, which are the only ones it needs to upload.

A manifest only consists of subsets, it's the component spec that lists the output subsets, so I guess we need that to check which ones to upload. So then we're checking output subsets twice? One time in the `evolve` method and one time in the `Dataset` class

Reply:
Sorry, you're right. The output subsets should be checked in the component spec, which the `Dataset` also needs to define which data to load.

So a `Dataset` can either
- load data based on an input manifest and a component specification
- write data based on an output manifest and a component specification

To be seen if it's logical to keep both in a single class or if we should split them.

Reply:
Closing this discussion as it's now incorporated